lib/observability/tracing/tracing_test.go

```diff
@@ -88,144 +79,53 @@ func generateTLSCertificate() (tls.Certificate, error) {
-
-	coltracepb.TraceServiceServer
-	spans []*otlp.ScopeSpans
-}
-
-type collectorConfig struct {
-	WithTLS bool
-}
-
-func newCollector(cfg collectorConfig) (*collector, error) {
-	grpcLn, err := net.Listen("tcp4", "")
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
-
-	httpLn, err := net.Listen("tcp4", "")
-	if err != nil {
-		return nil, trace.Wrap(err)
-	}
-
-	var tlsConfig *tls.Config
-	creds := insecure.NewCredentials()
-	if cfg.WithTLS {
-		tlsCertificate, err := generateTLSCertificate()
-		if err != nil {
-			return nil, err
-		}
-		tlsConfig = &tls.Config{
-			Certificates: []tls.Certificate{tlsCertificate},
-		}
-		creds = credentials.NewTLS(tlsConfig)
-	}
-
-	c := &collector{
-		grpcLn:     grpcLn,
-		httpLn:     httpLn,
-		grpcServer: grpc.NewServer(grpc.Creds(creds)),
-	}
-
-	c.httpServer = &http.Server{Handler: c, TLSConfig: tlsConfig}
-
-	coltracepb.RegisterTraceServiceServer(c.grpcServer, c)
-
-	return c, nil
-}
-
-func (c *collector) ClientTLSConfig() *tls.Config {
-	if c.httpServer.TLSConfig == nil {
-		return nil
-	}
+func TestNewClient(t *testing.T) {
+	t.Parallel()
+	c, err := NewCollector(CollectorConfig{})
+	require.NoError(t, err)
 
-	return &tls.Config{
-		InsecureSkipVerify: true,
+	cfg := Config{
+		Service:     "test",
+		ExporterURL: c.GRPCAddr(),
+		DialTimeout: time.Second,
 	}
-}
-
-func (c collector) GRPCAddr() string {
-	return "grpc://" + c.grpcLn.Addr().String()
-}
-
-func (c collector) HTTPAddr() string {
-	return "http://" + c.httpLn.Addr().String()
-}
-
-func (c collector) HTTPSAddr() string {
-	return "https://" + c.httpLn.Addr().String()
-}
-
-func (c collector) Start() error {
-	var g errgroup.Group
-
-	g.Go(func() error {
-		if c.httpServer.TLSConfig != nil {
-			return c.httpServer.ServeTLS(c.httpLn, "", "")
-		}
-		return c.httpServer.Serve(c.httpLn)
-	})
 
-	g.Go(func() error {
-		return c.grpcServer.Serve(c.grpcLn)
+	t.Cleanup(func() {
+		shutdownCtx, shutdownCancel := context.WithTimeout(context.Background(), 3*time.Second)
+		defer shutdownCancel()
+		require.NoError(t, c.Shutdown(shutdownCtx))
 	})
+	go func() {
+		c.Start()
+	}()
 
-	return g.Wait()
-}
-
-func (c collector) Shutdown(ctx context.Context) error {
-	c.grpcServer.Stop()
-	return c.httpServer.Shutdown(ctx)
-}
-
-func (c *collector) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	if r.URL.Path == "/v1/traces" {
-		response := coltracepb.ExportTraceServiceResponse{}
-		rawResponse, err := proto.Marshal(&response)
-		if err != nil {
-			w.WriteHeader(http.StatusInternalServerError)
-			return
-		}
-
-		rawRequest, err := io.ReadAll(r.Body)
-		if err != nil {
-			w.WriteHeader(http.StatusInternalServerError)
-			return
-		}
-
-		req := &coltracepb.ExportTraceServiceRequest{}
-		if err := proto.Unmarshal(rawRequest, req); err != nil {
-			w.WriteHeader(http.StatusBadRequest)
-			return
-		}
+	// NewClient shouldn't fail - it won't attempt to connect to the Collector
+	clt, err := NewClient(cfg)
+	require.NoError(t, err)
```

Comment:
I feel like those comments should be part of the error message when the test fails
But you did this so methodically that I'm confident it's on purpose
Can you please help me understand the rationale behind this pattern?